src/backend/distributed/commands/utility_hook.c

```diff
@@ -1626,18 +1663,96 @@ RunPreprocessMainDBCommand(Node *parsetree, const char *queryString)
  */
 static void
 RunPostprocessMainDBCommand(Node *parsetree)
+{
+	if (IsStatementSupportedInNonMainDb(parsetree) &&
+		StatementRequiresMarkDistributedFromNonMainDb(parsetree))
+	{
+		MarkObjectDistributedInNonMainDb(parsetree);
+	}
+}
+
+
+/*
+ * GetObjectInfo returns ObjectInfo for the target object of given parsetree.
+ */
+static ObjectInfo
+GetObjectInfo(Node *parsetree)
 {
 	if (IsA(parsetree, CreateRoleStmt))
 	{
-		StringInfo mainDBQuery = makeStringInfo();
-		CreateRoleStmt *createRoleStmt = castNode(CreateRoleStmt, parsetree);
-		Oid roleOid = get_role_oid(createRoleStmt->role, false);
-		appendStringInfo(mainDBQuery,
-						 MARK_OBJECT_DISTRIBUTED,
-						 AuthIdRelationId,
-						 quote_literal_cstr(createRoleStmt->role),
-						 roleOid,
-						 quote_literal_cstr(CurrentUserName()));
-		RunCitusMainDBQuery(mainDBQuery->data);
+		CreateRoleStmt *stmt = castNode(CreateRoleStmt, parsetree);
+		ObjectInfo info = {
+			.name = stmt->role,
+			.id = get_role_oid(stmt->role, false)
+		};
+
+		return info;
+	}
+
+	/* Add else if branches for other statement types */
+
+	elog(ERROR, "unsupported statement type");
+}
+
+
+/*
+ * MarkObjectDistributedInNonMainDb marks the given object as distributed on the
+ * non-main database.
+ */
+static void
+MarkObjectDistributedInNonMainDb(Node *parsetree)
+{
+	ObjectInfo objectInfo = GetObjectInfo(parsetree);
+	StringInfo mainDBQuery = makeStringInfo();
+	appendStringInfo(mainDBQuery,
+					 MARK_OBJECT_DISTRIBUTED,
+					 AuthIdRelationId,
+					 quote_literal_cstr(objectInfo.name),
+					 objectInfo.id,
+					 quote_literal_cstr(CurrentUserName()));
+	RunCitusMainDBQuery(mainDBQuery->data);
+}
+
+
+/*
+ * IsStatementSupportedInNonMainDb returns true if the statement is supported from a
+ * non-main database.
+ */
+static bool
+IsStatementSupportedInNonMainDb(Node *parsetree)
```

Comment:
nit: Given that we used `FromNonMainDb` when naming StatementRequiresMarkDistributedFromNonMainDb, can we use the same here?

IsStatementSupportedInNonMainDb -> IsStatementSupportedFromNonMainDb

Reply:
> nit: Given that we used ..FromNonMainDb wording when naming StatementRequiresMarkDistributedFromNonMainDb, can we use the same here?

IsStatementSupportedInNonMainDb -> IsStatementSupportedFromNonMainDb

I believe we should also rename MarkObjectDistributedOnNonMainDb to MarkObjectDistributedFromNonMainDb now, given above reasoning.

Reply:
> nit: Given that we used ..FromNonMainDb wording when naming StatementRequiresMarkDistributedFromNonMainDb, can we use the same here?
>
> IsStatementSupportedInNonMainDb -> IsStatementSupportedFromNonMainDb

I believe we should also rename MarkObjectDistributedOnNonMainDb to MarkObjectDistributedFromNonMainDb now, given above reasoning.

Reply:
Renamed